---
title: "problemSet2-DESeq2-aedavids.Rmd"
output: html_notebook
---

Questions 2) Use DESeq2 to perform differential gene expression analysis comparing control replicates versus HOXA1 KD replicates. 

ref:
-[https://bioconductor.org/packages/release/bioc/html/DESeq2.html](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)
- [DESeq2 vignettes](http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pre-filtering)
- [DESeqDataSetFromHTSeqCount](https://www.rdocumentation.org/packages/DESeq/versions/1.24.0/topics/DESeqDataSet-class)

```{r}
library("DESeq2")
```

```{r}
directory <- "/Users/andrewdavidson/googleUCSC/bme-237-appliedRNABioinformatics/homework/problemSet2/data/reads/htseq-counts.out"
sampleFilesList <- list.files(directory)
sampleFilesList
```

```{r}
# strsplit(d[[1]], "\\.")
# sampleName <- strsplit(d[[1]], "\\.")[[1]][2]
# print( sampleName )
# treatment <- strsplit(d[[1]], "\\.")[[1]][3]
# print( treatment)
```

## create the sampleTable. 
you can add additional columns if you have more meta data
```{r}
# pre allocate list, this is faster than append
n <- length(sampleFilesList)
sampleNamesVec <- vector("list", n)
conditionVec  <- vector("list", n)

i <- 1
for (fileName in sampleFilesList) {
  #print( fileName)
  
  sampleName <- strsplit(fileName[[1]], "\\.")[[1]][3]
  #print( sampleName )
  #sampleNamesList <- c(sampleNamesList, sampleName)
  sampleNamesVec[[i]] <- sampleName
  
  condition <- strsplit(fileName[[1]], "\\.")[[1]][4]
  #print(condition)
  #conditionList <- c(conditionList, condtion)
  conditionVec[[i]] <- condition
  
  i <- i + 1
}

# flatten lists
sampleNamesVec <- unlist(sampleNamesVec)
conditionVec <- unlist(conditionVec)
print( sampleNamesVec )
print(conditionVec )
```

```{r}
sampleTable <- data.frame(sampleName = sampleNamesVec,
                          fileName = sampleFilesList,
                          condition = conditionVec)
sampleTable$condition <- factor(sampleTable$condition)

sampleTable
```


## create the DESeqDataSet
```{r}
ddsHTSeq <- DESeqDataSetFromHTSeqCount(sampleTable, 
                                  directory, 
                                  design = ~ condition)
ddsHTSeq
```

## data prep pre-filtering 
While it is not necessary to pre-filter low count genes before running the DESeq2 functions, there are two reasons which make pre-filtering useful: by removing rows in which there are very few reads, we reduce the memory size of the dds data object, and we increase the speed of the transformation and testing functions within DESeq2. Here we perform a minimal pre-filtering to keep only rows that have at least 10 reads total. Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function.

```{r}
keep <- rowSums(counts(ddsHTSeq)) >= 10
dds <- ddsHTSeq[keep,]
```


## relevel condition
 if you never tell the DESeq2 functions which level you want to compare against (e.g. which level represents the control group), the comparisons will be based on the alphabetical order of the levels.
 
```{r}
refference <- "control"
dds$condition <- relevel(dds$condition, ref = refference)
```


## Differential expression analysis

