{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BME-237 Applied RNA Bioinformatics problem set 1\n",
    "aedavids@ucsc.edu\n",
    "april 2021"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question 11\n",
    "Find a splice junction that was detected by STAR, but not detected by HISAT2.\n",
    "\n",
    "What is the chromosome, start position, and end position of the intron that corresponds to this splice junction?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Icon?\r\n",
      "hisat2.ENCFF000IJ.junctions\r\n",
      "hisat2.ENCFF000IJ.junctions.geneIds\r\n",
      "hisat2.ENCFF000IJ.junctions.jcounts\r\n",
      "hisat2.ENCFF000IJ.junctions.jcounts.primaryGene\r\n",
      "hisat2.ENCFF000IJ.junctions.summary\r\n",
      "jcount.primaryGenes\r\n",
      "junctions.tsv\r\n",
      "q18.GL000194.1.3reads.txt\r\n",
      "star.ENCFF000IJ.junctions\r\n",
      "star.ENCFF000IJ.junctions.geneIds\r\n",
      "star.ENCFF000IJ.junctions.jcounts\r\n",
      "star.ENCFF000IJ.junctions.jcounts.primaryGenes\r\n",
      "star.ENCFF000IJ.junctions.summary\r\n"
     ]
    }
   ],
   "source": [
    "!ls featureCounts.out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "starJunctions   = \"featureCounts.out/star.ENCFF000IJ.junctions\"\n",
    "hisat2Junctions = \"featureCounts.out/hisat2.ENCFF000IJ.junctions\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "starJuncDF   = pd.read_csv( starJunctions, sep='\\t', skiprows = 1 )\n",
    "starJuncDF.rename(columns={ starJuncDF.columns[-1]: \"count\" }, inplace = True)\n",
    "\n",
    "hisat2JuncDF = pd.read_csv( hisat2Junctions, sep='\\t', skiprows = 1 )\n",
    "hisat2JuncDF.rename(columns={ hisat2JuncDF.columns[-1]: \"count\" }, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s-Geneid</th>\n",
       "      <th>s-Chr</th>\n",
       "      <th>s-Start</th>\n",
       "      <th>s-End</th>\n",
       "      <th>s-Strand</th>\n",
       "      <th>s-Length</th>\n",
       "      <th>s-count</th>\n",
       "      <th>h-Geneid</th>\n",
       "      <th>h-Chr</th>\n",
       "      <th>h-Start</th>\n",
       "      <th>h-End</th>\n",
       "      <th>h-Strand</th>\n",
       "      <th>h-Length</th>\n",
       "      <th>h-count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1</td>\n",
       "      <td>11869;12010;12179;12613;12613;12975;13221;1322...</td>\n",
       "      <td>12227;12057;12227;12721;12697;13052;14409;1337...</td>\n",
       "      <td>+;+;+;+;+;+;+;+;+</td>\n",
       "      <td>1735</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1</td>\n",
       "      <td>11869;12010;12179;12613;12613;12975;13221;1322...</td>\n",
       "      <td>12227;12057;12227;12721;12697;13052;14409;1337...</td>\n",
       "      <td>+;+;+;+;+;+;+;+;+</td>\n",
       "      <td>1735</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...</td>\n",
       "      <td>14404;15005;15796;16607;16858;17233;17606;1791...</td>\n",
       "      <td>14501;15038;15947;16765;17055;17368;17742;1806...</td>\n",
       "      <td>-;-;-;-;-;-;-;-;-;-;-</td>\n",
       "      <td>1351</td>\n",
       "      <td>40</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...</td>\n",
       "      <td>14404;15005;15796;16607;16858;17233;17606;1791...</td>\n",
       "      <td>14501;15038;15947;16765;17055;17368;17742;1806...</td>\n",
       "      <td>-;-;-;-;-;-;-;-;-;-;-</td>\n",
       "      <td>1351</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1</td>\n",
       "      <td>29554;30267;30564;30976;30976</td>\n",
       "      <td>30039;30667;30667;31109;31097</td>\n",
       "      <td>+;+;+;+;+</td>\n",
       "      <td>1021</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1</td>\n",
       "      <td>29554;30267;30564;30976;30976</td>\n",
       "      <td>30039;30667;30667;31109;31097</td>\n",
       "      <td>+;+;+;+;+</td>\n",
       "      <td>1021</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>138</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>138</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            s-Geneid                                              s-Chr  \\\n",
       "0  ENSG00000223972.5       chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1   \n",
       "1  ENSG00000227232.5  chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...   \n",
       "2  ENSG00000278267.1                                               chr1   \n",
       "3  ENSG00000243485.5                           chr1;chr1;chr1;chr1;chr1   \n",
       "4  ENSG00000284332.1                                               chr1   \n",
       "\n",
       "                                             s-Start  \\\n",
       "0  11869;12010;12179;12613;12613;12975;13221;1322...   \n",
       "1  14404;15005;15796;16607;16858;17233;17606;1791...   \n",
       "2                                              17369   \n",
       "3                      29554;30267;30564;30976;30976   \n",
       "4                                              30366   \n",
       "\n",
       "                                               s-End               s-Strand  \\\n",
       "0  12227;12057;12227;12721;12697;13052;14409;1337...      +;+;+;+;+;+;+;+;+   \n",
       "1  14501;15038;15947;16765;17055;17368;17742;1806...  -;-;-;-;-;-;-;-;-;-;-   \n",
       "2                                              17436                      -   \n",
       "3                      30039;30667;30667;31109;31097              +;+;+;+;+   \n",
       "4                                              30503                      +   \n",
       "\n",
       "   s-Length  s-count           h-Geneid  \\\n",
       "0      1735        0  ENSG00000223972.5   \n",
       "1      1351       40  ENSG00000227232.5   \n",
       "2        68        0  ENSG00000278267.1   \n",
       "3      1021        0  ENSG00000243485.5   \n",
       "4       138        0  ENSG00000284332.1   \n",
       "\n",
       "                                               h-Chr  \\\n",
       "0       chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1   \n",
       "1  chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...   \n",
       "2                                               chr1   \n",
       "3                           chr1;chr1;chr1;chr1;chr1   \n",
       "4                                               chr1   \n",
       "\n",
       "                                             h-Start  \\\n",
       "0  11869;12010;12179;12613;12613;12975;13221;1322...   \n",
       "1  14404;15005;15796;16607;16858;17233;17606;1791...   \n",
       "2                                              17369   \n",
       "3                      29554;30267;30564;30976;30976   \n",
       "4                                              30366   \n",
       "\n",
       "                                               h-End               h-Strand  \\\n",
       "0  12227;12057;12227;12721;12697;13052;14409;1337...      +;+;+;+;+;+;+;+;+   \n",
       "1  14501;15038;15947;16765;17055;17368;17742;1806...  -;-;-;-;-;-;-;-;-;-;-   \n",
       "2                                              17436                      -   \n",
       "3                      30039;30667;30667;31109;31097              +;+;+;+;+   \n",
       "4                                              30503                      +   \n",
       "\n",
       "   h-Length  h-count  \n",
       "0      1735        0  \n",
       "1      1351       59  \n",
       "2        68        0  \n",
       "3      1021        0  \n",
       "4       138        0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc = [\"s-\" + c for c in starJuncDF]\n",
    "hc = [\"h-\" + c for c in hisat2JuncDF]\n",
    "df = pd.concat( [starJuncDF, hisat2JuncDF], axis=1)\n",
    "df.columns = sc + hc\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s-Geneid</th>\n",
       "      <th>s-Chr</th>\n",
       "      <th>s-Start</th>\n",
       "      <th>s-End</th>\n",
       "      <th>s-Strand</th>\n",
       "      <th>s-Length</th>\n",
       "      <th>s-count</th>\n",
       "      <th>h-Geneid</th>\n",
       "      <th>h-Chr</th>\n",
       "      <th>h-Start</th>\n",
       "      <th>h-End</th>\n",
       "      <th>h-Strand</th>\n",
       "      <th>h-Length</th>\n",
       "      <th>h-count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...</td>\n",
       "      <td>14404;15005;15796;16607;16858;17233;17606;1791...</td>\n",
       "      <td>14501;15038;15947;16765;17055;17368;17742;1806...</td>\n",
       "      <td>-;-;-;-;-;-;-;-;-;-;-</td>\n",
       "      <td>1351</td>\n",
       "      <td>40</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...</td>\n",
       "      <td>14404;15005;15796;16607;16858;17233;17606;1791...</td>\n",
       "      <td>14501;15038;15947;16765;17055;17368;17742;1806...</td>\n",
       "      <td>-;-;-;-;-;-;-;-;-;-;-</td>\n",
       "      <td>1351</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...</td>\n",
       "      <td>89295;92091;92230;110953;112700;112700;112700;...</td>\n",
       "      <td>91629;92240;92240;111357;112804;112804;112804;...</td>\n",
       "      <td>-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-</td>\n",
       "      <td>3726</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...</td>\n",
       "      <td>89295;92091;92230;110953;112700;112700;112700;...</td>\n",
       "      <td>91629;92240;92240;111357;112804;112804;112804;...</td>\n",
       "      <td>-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-</td>\n",
       "      <td>3726</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSG00000279928.2</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1</td>\n",
       "      <td>182696;183132;183494;183740;183981</td>\n",
       "      <td>182746;183216;183571;183901;184174</td>\n",
       "      <td>+;+;+;+;+</td>\n",
       "      <td>570</td>\n",
       "      <td>119</td>\n",
       "      <td>ENSG00000279928.2</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1</td>\n",
       "      <td>182696;183132;183494;183740;183981</td>\n",
       "      <td>182746;183216;183571;183901;184174</td>\n",
       "      <td>+;+;+;+;+</td>\n",
       "      <td>570</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1</td>\n",
       "      <td>185217;185491;186317;187129;187376;187755;1881...</td>\n",
       "      <td>185350;185559;186469;187287;187577;187890;1882...</td>\n",
       "      <td>-;-;-;-;-;-;-;-;-;-</td>\n",
       "      <td>1397</td>\n",
       "      <td>44</td>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1</td>\n",
       "      <td>185217;185491;186317;187129;187376;187755;1881...</td>\n",
       "      <td>185350;185559;186469;187287;187577;187890;1882...</td>\n",
       "      <td>-;-;-;-;-;-;-;-;-;-</td>\n",
       "      <td>1397</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...</td>\n",
       "      <td>257864;257913;258144;258524;258568;261550;2630...</td>\n",
       "      <td>259025;259025;259025;259025;259024;261634;2682...</td>\n",
       "      <td>-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-</td>\n",
       "      <td>8224</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...</td>\n",
       "      <td>257864;257913;258144;258524;258568;261550;2630...</td>\n",
       "      <td>259025;259025;259025;259025;259024;261634;2682...</td>\n",
       "      <td>-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-</td>\n",
       "      <td>8224</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              s-Geneid                                              s-Chr  \\\n",
       "1    ENSG00000227232.5  chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...   \n",
       "9    ENSG00000238009.6  chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...   \n",
       "18   ENSG00000279928.2                           chr1;chr1;chr1;chr1;chr1   \n",
       "19   ENSG00000279457.4  chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1   \n",
       "21  ENSG00000228463.10  chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...   \n",
       "\n",
       "                                              s-Start  \\\n",
       "1   14404;15005;15796;16607;16858;17233;17606;1791...   \n",
       "9   89295;92091;92230;110953;112700;112700;112700;...   \n",
       "18                 182696;183132;183494;183740;183981   \n",
       "19  185217;185491;186317;187129;187376;187755;1881...   \n",
       "21  257864;257913;258144;258524;258568;261550;2630...   \n",
       "\n",
       "                                                s-End  \\\n",
       "1   14501;15038;15947;16765;17055;17368;17742;1806...   \n",
       "9   91629;92240;92240;111357;112804;112804;112804;...   \n",
       "18                 182746;183216;183571;183901;184174   \n",
       "19  185350;185559;186469;187287;187577;187890;1882...   \n",
       "21  259025;259025;259025;259025;259024;261634;2682...   \n",
       "\n",
       "                               s-Strand  s-Length  s-count  \\\n",
       "1                 -;-;-;-;-;-;-;-;-;-;-      1351       40   \n",
       "9     -;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-      3726        0   \n",
       "18                            +;+;+;+;+       570      119   \n",
       "19                  -;-;-;-;-;-;-;-;-;-      1397       44   \n",
       "21  -;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-      8224        0   \n",
       "\n",
       "              h-Geneid                                              h-Chr  \\\n",
       "1    ENSG00000227232.5  chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...   \n",
       "9    ENSG00000238009.6  chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...   \n",
       "18   ENSG00000279928.2                           chr1;chr1;chr1;chr1;chr1   \n",
       "19   ENSG00000279457.4  chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1   \n",
       "21  ENSG00000228463.10  chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...   \n",
       "\n",
       "                                              h-Start  \\\n",
       "1   14404;15005;15796;16607;16858;17233;17606;1791...   \n",
       "9   89295;92091;92230;110953;112700;112700;112700;...   \n",
       "18                 182696;183132;183494;183740;183981   \n",
       "19  185217;185491;186317;187129;187376;187755;1881...   \n",
       "21  257864;257913;258144;258524;258568;261550;2630...   \n",
       "\n",
       "                                                h-End  \\\n",
       "1   14501;15038;15947;16765;17055;17368;17742;1806...   \n",
       "9   91629;92240;92240;111357;112804;112804;112804;...   \n",
       "18                 182746;183216;183571;183901;184174   \n",
       "19  185350;185559;186469;187287;187577;187890;1882...   \n",
       "21  259025;259025;259025;259025;259024;261634;2682...   \n",
       "\n",
       "                               h-Strand  h-Length  h-count  \n",
       "1                 -;-;-;-;-;-;-;-;-;-;-      1351       59  \n",
       "9     -;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-      3726        1  \n",
       "18                            +;+;+;+;+       570       96  \n",
       "19                  -;-;-;-;-;-;-;-;-;-      1397      106  \n",
       "21  -;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-      8224        2  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selectRowsWithReads = ( (df[\"s-count\"] != 0) | (df[\"h-count\"] != 0) )\n",
    "hasReadsDF = df.loc[ selectRowsWithReads, : ]\n",
    "hasReadsDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s-Geneid</th>\n",
       "      <th>s-count</th>\n",
       "      <th>s-Length</th>\n",
       "      <th>h-Geneid</th>\n",
       "      <th>h-count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>ENSG00000284740.1</td>\n",
       "      <td>1</td>\n",
       "      <td>660</td>\n",
       "      <td>ENSG00000284740.1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>ENSG00000269925.1</td>\n",
       "      <td>5</td>\n",
       "      <td>393</td>\n",
       "      <td>ENSG00000269925.1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>ENSG00000234546.4</td>\n",
       "      <td>16</td>\n",
       "      <td>2141</td>\n",
       "      <td>ENSG00000234546.4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>ENSG00000202415.1</td>\n",
       "      <td>2</td>\n",
       "      <td>319</td>\n",
       "      <td>ENSG00000202415.1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>ENSG00000201801.1</td>\n",
       "      <td>4</td>\n",
       "      <td>120</td>\n",
       "      <td>ENSG00000201801.1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              s-Geneid  s-count  s-Length           h-Geneid  h-count\n",
       "119  ENSG00000284740.1        1       660  ENSG00000284740.1        0\n",
       "275  ENSG00000269925.1        5       393  ENSG00000269925.1        0\n",
       "320  ENSG00000234546.4       16      2141  ENSG00000234546.4        0\n",
       "343  ENSG00000202415.1        2       319  ENSG00000202415.1        0\n",
       "411  ENSG00000201801.1        4       120  ENSG00000201801.1        0"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selectStarOnlyRows = ( (hasReadsDF[\"s-count\"] != 0) & (hasReadsDF[\"h-count\"] == 0) )\n",
    "starOnlyDF = hasReadsDF.loc[ selectStarOnlyRows ]\n",
    "starOnlyDF.loc[:,[\"s-Geneid\", \"s-count\", \"s-Length\", \"h-Geneid\", \"h-count\",]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s-Geneid</th>\n",
       "      <th>s-Chr</th>\n",
       "      <th>s-Start</th>\n",
       "      <th>s-End</th>\n",
       "      <th>s-Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>ENSG00000284740.1</td>\n",
       "      <td>chr1;chr1;chr1;chr1</td>\n",
       "      <td>1503250;1503582;1509205;1509205</td>\n",
       "      <td>1503661;1503661;1509452;1509351</td>\n",
       "      <td>660</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              s-Geneid                s-Chr                          s-Start  \\\n",
       "119  ENSG00000284740.1  chr1;chr1;chr1;chr1  1503250;1503582;1509205;1509205   \n",
       "\n",
       "                               s-End  s-Length  \n",
       "119  1503661;1503661;1509452;1509351       660  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "starOnlyDF.loc[ starOnlyDF[\"s-Geneid\"] == \"ENSG00000284740.1\", [\"s-Geneid\", \"s-Chr\", \"s-Start\", \"s-End\", \"s-Length\"] ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Q 16\n",
    "Find a splice junction that was detected by HISAT2, but not detected by STAR.\n",
    "\n",
    "What is the chromosome, start position, and end position of the intron that corresponds to this splice junction?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s-Geneid</th>\n",
       "      <th>s-count</th>\n",
       "      <th>h-Geneid</th>\n",
       "      <th>h-Length</th>\n",
       "      <th>h-count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>3726</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>8224</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENSG00000225630.1</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000225630.1</td>\n",
       "      <td>1044</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ENSG00000229344.1</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000229344.1</td>\n",
       "      <td>682</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>ENSG00000198744.5</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198744.5</td>\n",
       "      <td>547</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              s-Geneid  s-count            h-Geneid  h-Length  h-count\n",
       "9    ENSG00000238009.6        0   ENSG00000238009.6      3726        1\n",
       "21  ENSG00000228463.10        0  ENSG00000228463.10      8224        2\n",
       "34   ENSG00000225630.1        0   ENSG00000225630.1      1044        8\n",
       "37   ENSG00000229344.1        0   ENSG00000229344.1       682        8\n",
       "40   ENSG00000198744.5        0   ENSG00000198744.5       547        1"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selectHisat2OnlyRows = ( (hasReadsDF[\"h-count\"] != 0) & (hasReadsDF[\"s-count\"] == 0) )\n",
    "hisat2OnlyDF = hasReadsDF.loc[ selectHisat2OnlyRows ]\n",
    "hisat2OnlyDF.loc[:,[\"s-Geneid\", \"s-count\",  \"h-Geneid\", \"h-Length\", \"h-count\",]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>h-Geneid</th>\n",
       "      <th>h-Chr</th>\n",
       "      <th>h-Start</th>\n",
       "      <th>h-End</th>\n",
       "      <th>h-Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...</td>\n",
       "      <td>89295;92091;92230;110953;112700;112700;112700;...</td>\n",
       "      <td>91629;92240;92240;111357;112804;112804;112804;...</td>\n",
       "      <td>3726</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            h-Geneid                                              h-Chr  \\\n",
       "9  ENSG00000238009.6  chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...   \n",
       "\n",
       "                                             h-Start  \\\n",
       "9  89295;92091;92230;110953;112700;112700;112700;...   \n",
       "\n",
       "                                               h-End  h-Length  \n",
       "9  91629;92240;92240;111357;112804;112804;112804;...      3726  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hisat2OnlyDF.loc[ hisat2OnlyDF[\"h-Geneid\"] == \"ENSG00000238009.6\", [\"h-Geneid\", \"h-Chr\", \"h-Start\", \"h-End\", \"h-Length\"] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:extraCellularRNA] *",
   "language": "python",
   "name": "conda-env-extraCellularRNA-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
